MAINT: SciPy 1.14.1 backports

tylerjereddy opened the pull request:
Prep for release that supports CPython `3.13` now that NumPy `2.1.0rc1` is out.

Backports included:

1. gh-21062
2. gh-21067
3. gh-21069
4. gh-21084
5. gh-21107
6. gh-21108
7. gh-21160
8. gh-21209
9. gh-21274
10. gh-21283
11. gh-21315 (minus free-threading support)
12. gh-21324 (merge conflicts and only partial backport of patch)
13. gh-21320
14. gh-21200 (merge conflicts resolved)
15. gh-21244
16. gh-21271
17. gh-21104
18. gh-21400
19. gh-21172
20. gh-21074 (minor merge conflicts resolved)
21. gh-21000 (partial)
22. gh-21409
23. gh-21401 (**substantial** merge conflicts)

TODO:

- [x] get regular CI passing -- ~I saw 1 failure locally on ARM mac for `TestConstructUtils::test_vstack_1d_with_2d`, so I doubt regular CI will be happy yet~
- [x] get wheel builds working; experiments started at gh-21359
- [x] any other backports needed in general?
- [x] read diff carefully before merge
- [x] update release notes again after backport activity is done

steppi committed:
BUG: special: remove type punning (#21067)

to avoid warnings in LTO builds

steppi committed:
BUG: special: Fixes for pro_rad1 (#21062)

* BUG: Fix indices in sphj

* Fix sphj argument order

* Fix translation error from original Fortran

* TST: Add test for pro_rad1

* Bump test tolerance for linux 32bit

h-vetinari committed:
MAINT: fix typo in small_dynamic_array.h (#21069)

on clang-19, this causes:
```
../scipy/_lib/_uarray/small_dynamic_array.h(145,18): error: reference to non-static member function must be called
  145 |     size_ = copy.size;
      |             ~~~~~^~~~
1 error generated.
```
I'm not sure how previous versions (much less other compilers) dealt with this,
as it seems that the `SmallDynamicArray` class has no `size` member or field at all.

nickodell committed:
BLD: Enable open_memstream on newer glibc

On glibc >= 2.10, if -std=c17 is set and no feature test macros are
set, glibc will not define an open_memstream definition. musl has
similar behavior. This causes SciPy to unnecessarily open a temporary
file, when capturing the error stream could have been done in memory.

lucascolley committed:
MAINT: adapt to array-api-strict 2.0 (#21074)

* MAINT: adapt to array-api-strict 2.0
* DEV: enforce array-api-strict>=2.0

[skip cirrus] [skip circle]

dschult committed:
add release note for 1.14 re: sparse array iteration

[docs only]

czgdp1807 committed:
MAINT: Unskip `scipy.misc.test.test_decorator` for Python 3.13+ (#21104)

Closes gh-19572

[docs only]

dschult committed:
BUG: sparse: fix 1D for vstack/hstack and Improve 1D error msgs in h/v-stacks/kron/kronsum (#21108)

* fix 1D for vstack/hstack and Improve 1D err msg stacks & kron/kronsum

* improve error messages in kronsum

Co-authored-by: CJ Carey <perimosocordiae@gmail.com>

* adjust exception test match string

---------

Co-authored-by: CJ Carey <perimosocordiae@gmail.com>

rgommers committed:
BUG: signal: fix crash under free-threaded CPython in medfilt2d

This also improves the logic; the `setjmp`/`longjmp` nor the
`check_malloc` function were good ways to deal with error
handling when memory allocation failed.
It looks like that was introduced because the type-generic
`*_medfilt2d` defined through a macro cannot have an `int`
return type (only `void` for macros)`. However, an integer
error value can instead be passed to the function as an extra
argument.

Closes gh-21142

AA-Turner committed:
Fix type of ``html_sidebars`` value in ``conf.py``

melissawm committed:
DOC: Suppress autosummary module names warning

May be fixed by upstream in the future.

[docs only]

lucascolley committed:
BUG: fft: fix array-like input (#21209)

dschult committed:
BUG: sparse: fix failing doctests that print sparse (#21244)

* fix doctests using sparse in csgraph and io

* ensure no mixed spaces and tabs

[docs only]

ev-br committed:
DOC: stats: silence the doctest error

[docs only]

rgommers committed:
MAINT: sparse.linalg: update `SuperLU/colamd.c` to fix license issue (#21274)

This applies the upstream commit (e08261a7c9) that resolved a problem
with this file - it mentioned both BSD and LGPL licenses.

Closes gh-21234

EwoutH committed:
CI: Update to cibuildwheel 2.20.0
cibuildwheel 2.20.0 uses the ABI stable Python 3.13.0rc1 and build Python 3.13 wheels by default.

TiborVoelcker committed:
BUG: odr: fix pickling

The `__getattr__` method of the `odr` objects led to a recursion error
when unpickling. This is because, when unpickling, the instance is
checked for a `__setstate__` method, before the any attributes are
added.
The implementation before assumed that the attributes are already set.
Closes gh-21306.

TiborVoelcker committed:
STY: odr: Remove double newlines

WarrenWeckesser committed:
DOC: Don't use Sphinx 8.0.0 until gh-21323 is resolved.

[docs only]

lucascolley committed:
BUG: stats: adapt to `np.floor` type promotion removal (#21283)

* BUG: stats: adapt to `np.floor` type promotion removal

`rv_discrete._cdf` relied on `np.floor` promoting its integer input to `np.float64`. This is no longer the case since numpy/numpy#26766.

tylerjereddy committed:
MAINT, DOC: update 1.14.1 relnotes

* Update the SciPy `1.14.1` release notes
following backport activity, and to clarify
that this release add support for Python `3.13`.

* Also update `pyproject.toml` for `3.13` wheel
builds.

[skip cirrus]

dschmitz89 commented:
If it gets merged in time, we should also include [#update superlu](https://github.com/scipy/scipy/pull/21172).

EwoutH commented:
Some parts of is https://github.com/scipy/scipy/pull/20822 and https://github.com/scipy/scipy/pull/21104 need to be backported for Python 3.13 support, as wel as https://github.com/scipy/scipy/pull/21359/commits/f20a0b4572ab153b3ac3d2c16d7b523bf8c200d3.

What parts exactly depends on if we want to support free-threaded Python 3.13 (I think the conclusion was no).

tylerjereddy commented on scipy/sparse/tests/test_construct.py:
```diff
@@ -459,6 +475,19 @@ def test_vstack_matrix_or_array(self):
         assert isinstance(construct.vstack([coo_matrix(A), coo_array(B)]), sparray)
         assert isinstance(construct.vstack([coo_matrix(A), coo_matrix(B)]), spmatrix)
 
+    def test_vstack_1d_with_2d(self):
```
@perimosocordiae @dschult This test, backported from gh-21108, fails on the maintenance branch because 1D sparse array support is not as complete on the older branch. How would you like me to proceed?

If I delete that test, the full suite passes locally. Given this information, I'm not sure if you want me to do that or back the entire patch out, etc., so let me know what you prefer.

rgommers replied:
The reproducer for the regression reported in gh-21064 is not fixed in this branch, which is what gh-21108 aimed to do. So it looks to me like backing out the backport is better than deleting the failing test.

Of course if it's fixable by another backport that is even better. However, I don't think it's the end of the world if the backport is dropped, since it was a regression from 1.12.0 to 1.13.0 so that can't be fully repaired anyway - plus it's a pretty niche thing to do, so as regressions go it's not such a painful one.

dschult replied:
The changes in #21108 to vstack/hstack are more general than fixing only the reported regression. There may be other cases where folks want to input, e.g.  1D coo arrays into vstack, and the fix from #21108 avoids those problems. So I would prefer removing this test to removing the whole patch. BUT: see below where we add to the test and fix some similar bugs that should be backported too.

In working on this issue, I added a couple of other tests, some of which fail. I put those along with fixes in #21400. That PR should probably be backported with #21108 to allow it to both be tested more thoroughly and to pass those tests.

tylerjereddy committed:
DOC: sphinx upper bound

* Upper bounding the version of Sphinx used on the maintenance
branch seems to help avoid a doc build issue locally.

[docs only]

tylerjereddy commented:
Thanks, I added those and another docs backport that was needed. In addition, I also had to upper-bound the `sphinx` version more tightly on the release branch here to avoid other docs build errors locally. We'll see if CI agrees on docs build working or not. I made the upper bound decision fairly crudely--simply looking at `sphinx` PyPI release dates and comparing with `1.14.0` release date to guess last working version.

While doing that I ran `python tools/generate_requirements.py` and it made me start thinking about whether we should upper bound more broadly in the requirements files as well on the release branch--for now, I only preserved the pins I needed for the `doc.txt` one (those files are mostly for dev work anyway, so not sure it matters so much on the release branch).

Beyond that, I'm definitely a bit slow here--waiting for the `sparse` folks is just a convenient excuse, and I should be playing a bit more with `cibuildwheel` and `3.13` stuff locally.

tylerjereddy commented:
For refguide failures I think I'll also need gh-21244

rgommers commented:
> While doing that I ran `python tools/generate_requirements.py` and it made me start thinking about whether we should upper bound more broadly in the requirements files as well on the release branch--for now, I only preserved the pins I needed for the `doc.txt` one (those files are mostly for dev work anyway, so not sure it matters so much on the release branch).

A few pragmatic upper bounds, including on Sphinx major versions, could certainly be a good option. Sphinx is way too fragile, and even minor versions are often breaking.

It's not really fixable in a good way without a large structural change though. I'm thinking about those, and it's becoming more pressing over time as our dependency chains (both for docs and for other array libraries recently) become more complex. That's why I'm experimenting with this out of tree (xref https://github.com/scipy/scipy/issues/21291#issuecomment-2275083007).

tylerjereddy commented:
Still needs many more backports, even for the docs. Next up is gh-21271

tylerjereddy commented:
With Python `3.13.0rc1` on ARM Mac, this is branch is back to just 1 failure, the `sparse` one above. I'm going to ping in some other venues to get feedback on that tonight.

Also, @h-vetinari, any concerns on the `conda-forge` side wrt `1.14.1` and `3.13` (not free-threaded) support.

dschult committed:
BUG: sparse: Fix 1D specialty hstack codes (#21400)

* add more hstack/vstack tests to cover coo and csr special cases

* change shape to _shape_as_2d where needed in specialty block() functions.

[skip cirrus]

tylerjereddy commented:
`python dev.py smoke-tutorials` is failing on `doc/source/tutorial/stats/sampling.rst`, which was recently moved from RST to MyST-md format in gh-20303. The "failing" code block is identical in both, so I will probably go ahead and just add `# may vary` to the offending line -- fyi @mdhaber @melissawm let me know if you have objections.

tylerjereddy committed:
MAINT, DOC: PR 21362 revisions

* Two `stats` tutorial examples had
floating point expected values that could
vary, so allow for this variation on the
release branch to prevent `smoke-tutorials`
failures.

* One of the examples/reasons is explained here:
https://github.com/scipy/scipy/pull/21362#issuecomment-2292214328

[docs only]

tylerjereddy commented:
oy, now:

```
ERROR Missing dependencies:
	pybind11<2.13.0,>=2.12.0
```

I know `pybind11` did a very recent release, but that range looks fine. Sometimes it is a red herring message and it is another pin to think about..

tylerjereddy commented:
The full suite does pass locally so I think the `sparse` business has been resolved thanks to @dschult

tylerjereddy commented:
@lucascolley @mdhaber what is preferred course of action for array API CI fails: `The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12`?

tylerjereddy committed:
MAINT, CI: PR 21362 revisions

* Pin `pybind11` in the "oldest gcc" CI job
where a newer version was getting pulled in and
causing problems given the upper bound in our
`pyproject.toml`.

[skip cirrus]

tylerjereddy commented:
The `pybind11` thing was just needing to pin that in CI so we don't go above the bound in `pyproject.toml` (I think this happens on every release branch with that CI job now, over time).

For regular CI, I think the only two issues left are:

1) the known gh-21052 (also: gh-21364, same thing), which also fails on `main`, so can maybe ignore here if it doesn't interfere with wheel builds (pre-release NumPy only?)
2) The array API testing issue, `The function clip requires API version 2023.12 or later, but the current API version for array-api-strict is 2022.12`

dschmitz89 committed:
BUG: sparse.linalg: Update `SuperLU` to fix unfilterable output for singular matrices (#21172)

[ci skip]

lucascolley commented:
> what is preferred course of action for array API CI fails

Fails for the testing utility array-api-strict have no impact on end users, so we can just ignore them. Unless you want to make CI green, but I'm not sure what we would need to backport in that case.

tylerjereddy commented:
I backported your gh-21074, which seemed to help locally; we'll see if CI agrees on the array API job.

tylerjereddy commented:
ok, for regular CI is just the one job failing per https://github.com/scipy/scipy/pull/21362#issuecomment-2292467476 now, which is being addressed on `main` where it also occurs

tylerjereddy commented:
I'll start debugging the wheel builds on my fork at https://github.com/tylerjereddy/scipy/pull/79 and locally with `cibuildwheel` where possible. Warren is doing a good on the Windows int bugfix on `main` already.

tylerjereddy committed:
CI, BLD: 3.13 wheel changes [wheel build]

* Custom backport of gh-21000 to support Python `3.13`
wheel builds.

Co-authored-by: Andrew Nelson <andyfaff@gmail.com>

tylerjereddy committed:
MAINT, CI: wheel build changes [wheel build]

* Additional modified backporting from
gh-21000

Co-authored-by: Andrew Nelson <andyfaff@gmail.com>

tylerjereddy commented:
The fork debug wheel build PR is in pretty good shape now, so could be cherry-picked back to here soon.

 Just the Windows integer thing (https://github.com/scipy/scipy/pull/21401) and Cirrus 3.13 things (https://github.com/scipy/scipy/pull/21409) remain, I think.

andyfaff commented:
I think I might need to do some local debugging for the cirrus cp313 build.

andyfaff committed:
BLD: cp313 wheels on `manylinux_aarch64` (#21409)

WarrenWeckesser committed:
MAINT: special: Accommodate changed integer handling in NumPy 2.0. (#21401)

* Add the 'p' character to the typecodes recognized by the script
  `_generate_pyx.py`.  This code corresponds to the integer types
  `npy_intp` and `Py_ssize_t`.
* In Cython code:
  * In change the fused type `dl_number_t` to `dlp_number_t`, and
    include `Py_ssize_t` in the types it fuses.
  * Change other uses of `long` to `Py_ssize_t`.
* In the type signatures of ufuncs configured in `functions.json`,
  change occurrences of `i` or `l` in an input to `p`.
* Change use of `int` in the wrappers of `xsf` functions to `Py_ssize_t`.
  The thin wrappers cast this to `int` before calling the `xsf` function.

tylerjereddy committed:
DOC: update 1.14.1 relnotes [wheel build]

* Update the SciPy `1.14.1` release notes after
several more backports.

[wheel build]

tylerjereddy commented:
We're at parity with the current backport list, doc build/refguide/regular testsuite pass locally, rendered release notes docs look "ok" locally--proceeding with attempt at full wheel build here.

tylerjereddy committed:
MAINT, CI: PR 21362 revisions [wheel build]

* Try adding a workaround in the wheel builds
for gh-21416.

[wheel build]

tylerjereddy committed:
MAINT, BLD: PR 21362 revisions [wheel build]

* Try setting an upper bound to `setuptools` version
in `pyproject.toml` (this is probably terrible), to hack
around gh-21416.

[wheel build]

tylerjereddy committed:
MAINT, CI: PR 21362 revisions [wheel build]

* Restore build isolation for Python 3.13 wheel builds.

[wheel build]

tylerjereddy commented:
Because of gh-21416, I'll pause on this for now. I did push in a hack that worked locally to override the transitive broken `setuptools` dependency pulled in by `pythran`, by adding a dummy dependency on bounded `setuptools` for us directly in `pyproject.toml`, but shouldn't do that really of course. That said, the transitive dependency stuff is pretty nasty, even if we did disable build isolation the from-source experience would be broken without a hack like that I think, in the absence of an upstream fix/yank.

Hopefully that gets cleaned up upstream.

tylerjereddy committed:
MAINT: PR 21362 revisions [wheel build]

* Revert some commits related to `setuptools` breakage.
See gh-21416.

[wheel build]

tylerjereddy commented:
Reverted some of the commits related to `setuptools` shenanigans, and giving the wheel builds another shot.

EwoutH commented:
All green!

tylerjereddy commented:
I read over some of the more important files in the diff, and skimmed over the much larger patches just to make sure they matched my recollection, and nothing seems too weird.

With the full matrix, including wheel builds with 3.13, now green, I'll go ahead and merge and start the release process for `1.14.1`.

tylerjereddy commented:
Also, backport labels have been cleared/clean slate for next round of bugs.
